File: dart/leet_code/988.dart
/// Dart core library for fundamental classes.
import 'dart:core';

/// Dart implementation of the LeetCode problem 988: Smallest String Starting From Leaf.
///
/// This module contains a class `Solution` with a method `smallestFromLeaf`.
/// The `smallestFromLeaf` method takes in one parameter:
/// - `root`: A TreeNode object representing the root of the tree.
///
/// The method returns a string representing the smallest string that starts from leaf and one is smallest from all possible strings.
/// The letters in the string are from a to z only and not case sensitive.
/// You may assume that the input tree will always have at least one leaf and that its size won't exceed 8500.
///
/// Error Handling:
/// - This method assumes that the input parameter is well-formed, i.e., `root` is a TreeNode object.
/// - If the input parameter is not well-formed, the behavior of the method is undefined.
/// - Throws an `ArgumentError` if the root is null.

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;

  /// Constructor for the TreeNode class.
  ///
  /// @param val The value of the node.
  /// @param left The left child of the node.
  /// @param right The right child of the node.
  TreeNode([this.val = 0, this.left, this.right]);
}

class Solution {
  /// Finds the smallest string that starts from leaf and one is smallest from all possible strings.
  ///
  /// This method uses a depth-first search approach to traverse the tree and find the smallest string.
  /// It iterates over each node in the tree. If the node is a leaf node, it compares the current path with the smallest path found so far and updates it if necessary.
  /// It uses a string to keep track of the current path and appends the character corresponding to the current node's value to the beginning of the string.
  ///
  /// @param root The root of the tree.
  /// @return The smallest string that starts from leaf and one is smallest from all possible strings.
  String smallestFromLeaf(TreeNode? root) {
    String res = '{';
    void dfs(TreeNode? node, String path) {
      if (node == null) return;
      path = '${String.fromCharCode(97 + node.val)}$path';
      if (node.left == null && node.right == null) {
        res = path.compareTo(res) < 0 ? path : res;
      }
      dfs(node.left, path);
      dfs(node.right, path);
    }

    dfs(root, '');
    return res;
  }
}

File: typescript/leet_code/988.js
/**
 * @fileoverview This module provides a class to represent a tree node and a function to find the smallest string
 *     starting from leaf.
 * @module leet_code/988
 */
/**
 * Class representing a tree node.
 */
const TreeNode = /** @class */ (function() {
  /**
     * Create a tree node.
     * @param {number} val - The value of the node.
     * @param {TreeNode | null} left - The left child of the node.
     * @param {TreeNode | null} right - The right child of the node.
     */
  function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
  }

  return TreeNode;
}());

/**
 * Function to find the smallest string starting from leaf.
 * The function uses a depth-first search approach to traverse the tree.
 *
 * @param {TreeNode | null} root - The root of the tree.
 * @return {string} - The smallest string starting from leaf.
 * @throws {Error} - Throws an error if the input parameter is not valid.
 */
function smallestFromLeaf(root) {
  // Check if the input parameter is valid
  if (root === null) {
    throw new Error('Invalid input parameter');
  }
  let answer = '';

  /**
     * Helper function to perform depth-first search on the tree.
     *
     * @param {TreeNode} node - The current node.
     * @param {string} str - The current string.
     */
  function dfs(node, str) {
    const char = String.fromCharCode(node.val + 'a'.charCodeAt(0));
    // If the node is a leaf, update the answer if necessary
    if (node.left === null && node.right === null) {
      if (answer === '' || char + str < answer) {
        answer = char + str;
      }
      return;
    }
    // Recurse on the left child if it exists
    if (node.left) {
      dfs(node.left, char + str);
    }
    // Recurse on the right child if it exists
    if (node.right) {
      dfs(node.right, char + str);
    }
  }

  // Start the depth-first search from the root
  dfs(root, '');
  return answer;
}
